cvat/apps/engine/cache.py

```diff
@@ -65,63 +70,166 @@
+        cache_item_ttl: int = None,
+        **kwargs: Any,
     ) -> _CacheItem:
-        def create_item() -> _CacheItem:
-            slogger.glob.info(f"Starting to prepare chunk: key {key}")
-            item_data = create_callback()
-            slogger.glob.info(f"Ending to prepare chunk: key {key}")
+        item = self._get_cache_item(key)
+        if item:
+            return item
 
-            item_data_bytes = item_data[0].getvalue()
-            item = (item_data[0], item_data[1], self._get_checksum(item_data_bytes))
-            if item_data_bytes:
-                self._cache.set(key, item)
+        return self._create_cache_item(
+            key,
+            create_callback,
+            *args,
+            cache_item_ttl=cache_item_ttl,
+            **kwargs,
+        )
 
-            return item
+    def _get_queue(self) -> rq.Queue:
+        return django_rq.get_queue(self._QUEUE_NAME)
 
-        item = self._get_cache_item(key)
-        if not item:
-            item = create_item()
+    def _make_queue_job_id_base(self, key: str) -> str:
+        return f"{self._QUEUE_JOB_PREFIX_TASK}{key}"
+
+    @staticmethod
+    def _drop_return_value(func: Callable[..., DataWithMime], *args: Any, **kwargs: Any):
+        func(*args, **kwargs)
+
+    @classmethod
+    def _create_and_set_cache_item(
+        cls,
+        key: str,
+        create_callback: Callable[..., DataWithMime],
+        *args: Any,
+        cache_item_ttl: int = None,
+        **kwargs: Any,
+    ) -> DataWithMime:
+        item_data = create_callback(*args, **kwargs)
+        item_data_bytes = item_data[0].getvalue()
+        item = (item_data[0], item_data[1], cls._get_checksum(item_data_bytes))
+        if item_data_bytes:
+            cache = cls._cache()
+            cache.set(key, item, timeout=cache_item_ttl or cache.default_timeout)
+
+        return item
+
+    def _wait_for_rq_job(self, rq_job: rq.job.Job) -> bool:
+        retries = self._CHUNK_CREATE_TIMEOUT // self._SLEEP_TIMEOUT or 1
+        while retries > 0:
+            job_status = rq_job.get_status()
+            if job_status in ("finished",):
+                return True
+            elif job_status in ("failed",):
+                job_meta = rq_job.get_meta()
+                exc_type = job_meta.get(RQJobMetaField.EXCEPTION_TYPE, Exception)
+                exc_args = job_meta.get(RQJobMetaField.EXCEPTION_ARGS, ("Cannot create chunk",))
+                raise exc_type(*exc_args)
+
+            time.sleep(self._SLEEP_TIMEOUT)
+            retries -= 1
+
+        return False
+
+    def _create_cache_item(
+            self,
+            key: str,
+            create_callback: Callable[..., DataWithMime],
+            *args: Any,
+            cache_item_ttl: int = None,
+            **kwargs: Any,
+        ) -> _CacheItem:
+
+        slogger.glob.info(f"Starting to prepare chunk: key {key}")
+        if self._is_run_inside_rq():
+            item = self._create_and_set_cache_item(
```

Comment:
If you want to prepare chunks in parallel, consider adding a lock. If you want to go lock-free, consider the version parameter of `.set`.